packages/client/hmi-client/src/workflow/ops/model-config/tera-model-config.vue

```diff
@@ -143,16 +180,134 @@ const emit = defineEmits([
 	'close'
 ]);
 
+interface BasicKnobs {
+	name: string;
+	description: string;
+	initials: Initial[];
+	parameters: ModelParameter[];
+	timeseries: { [index: string]: any };
+}
+
+const knobs = ref<BasicKnobs>({
+	name: '',
+	description: '',
+	initials: [],
+	parameters: [],
+	timeseries: {}
+});
+
+const isSaveDisabled = computed(() => {
+	if (knobs.value.name === '') return true;
+	return false;
+});
+
+const kernelManager = new KernelSessionManager();
+let editor: VAceEditorInstance['_editor'] | null;
+const buildJupyterContext = () => {
+	const contextId = selectedConfigId.value ?? props.node.state.tempConfigId;
+	if (!model.value) {
+		logger.warn('Cannot build Jupyter context without a model');
+		return null;
+	}
+	return {
+		context: 'mira_config_edit',
+		language: 'python3',
+		context_info: {
+			id: contextId
+		}
+	};
+};
+const codeText = ref(
+	'# This environment contains the variable "model_config" to be read and updated'
+);
+const notebookResponse = ref();
+const sampleAgentQuestions = [
+	'What are the current parameters values?',
+	'update the parameters {gamma: 0.13}'
+];
+
+const appendCode = (data: any, property: string, runUpdatedCode = false) => {
+	codeText.value = codeText.value.concat(' \n', data.content[property] as string);
+	if (runUpdatedCode) runFromCode();
+};
+
+const runFromCode = () => {
+	const code = editor?.getValue();
+	if (!code) return;
+
+	const messageContent = {
+		silent: false,
+		store_history: false,
+		user_expressions: {},
+		allow_stdin: true,
+		stop_on_error: false,
+		code
+	};
+
+	let executedCode = '';
+
+	kernelManager
+		.sendMessage('execute_request', messageContent)
+		.register('execute_input', (data) => {
+			executedCode = data.content.code;
+		})
+		.register('stream', (data) => {
+			notebookResponse.value = data.content.text;
+		})
+		.register('error', (data) => {
+			logger.error(`${data.content.ename}: ${data.content.evalue}`);
+			console.log('error', data.content);
+		})
+		.register('model_preview', (data) => {
+			if (!data.content) return;
+			handleModelPreview(data);
+
+			if (executedCode) {
+				saveCodeToState(executedCode, true);
+			}
+		});
+};
+
+const saveCodeToState = (code: string, hasCodeBeenRun: boolean) => {
+	const state = _.cloneDeep(props.node.state);
+	state.hasCodeBeenRun = hasCodeBeenRun;
+
+	// for now only save the last code executed, may want to save all code executed in the future
+	const codeHistoryLength = props.node.state.modelEditCodeHistory.length;
+	const timestamp = Date.now();
+	if (codeHistoryLength > 0) {
+		state.modelEditCodeHistory[0] = { code, timestamp };
+	} else {
+		state.modelEditCodeHistory.push({ code, timestamp });
+	}
+	emit('update-state', state);
+};
```

Comment:
Q: is this pattern used everywhere we have this agent/notebook pairing?

Reply:
yes I believe so. This can be factored when we have some time